src/test/pages/TeacherRequest.test.tsx

```diff
@@ -0,0 +1,204 @@
+
+describe("TeacherRequest", () => {
+  const setupTest = ({
+    role,
+    extraEndpoints,
+    modifyUser,
+  }: {
+    role: TestUserRole;
+    extraEndpoints?: RestHandler<never>[];
+    modifyUser?: (u: typeof mockUser) => typeof mockUser;
+  }) => {
+    renderTestEnvironment({
+      role: role,
+      modifyUser: modifyUser,
+      PageComponent: TeacherRequest,
+      initialRouteEntries: ["/teacher_account_request"],
+      extraEndpoints: extraEndpoints,
+    });
+  };
+
+  it("renders without crashing", () => {
+    setupTest({ role: "STUDENT" });
+    checkPageTitle("Teacher account request");
+  });
+
+  it("displays a warning message if one is returned from the server", async () => {
+    const warningMessage = {
+      id: "teacher_registration_warning_message",
+      type: "isaacPageFragment",
+      canonicalSourceFile: "content/general_pages/fragments/teacher_registration_warning_message.json",
+      title: "Teacher registration",
+      encoding: "markdown",
+      children: [
+        {
+          type: "content",
+          encoding: "html",
+          children: [],
+          value: "<p>Test warning message</p>",
+          published: false,
+          tags: [],
+        },
+      ],
+    };
+    const extraEndpoints = [
+      rest.get(API_PATH + `/pages/fragments/teacher_registration_warning_message`, (req, res, ctx) => {
+        return res(ctx.status(200), ctx.json(warningMessage));
+      }),
+    ];
+    setupTest({ role: "STUDENT", extraEndpoints });
+    const warning = await screen.findByTestId("teacher-upgrade-warning");
+    expect(warning).toHaveTextContent("Test warning message");
+  });
+
+  it("does not display a warning message if none are returned from the server", () => {
+    const extraEndpoints = [
+      rest.get(API_PATH + `/pages/fragments/teacher_registration_warning_message`, (req, res, ctx) => {
+        return res(
+          ctx.status(404),
+          ctx.json({
+            responseCode: 404,
+            responseCodeType: "Not Found",
+            errorMessage:
+              "No content found that matches the query with parameters: [[teacher_registration_warning_message], [isaacPageFragment]]",
+            bypassGenericSiteErrorPage: false,
+          }),
+        );
+      }),
+    ];
+    setupTest({ role: "STUDENT", extraEndpoints });
+    const warning = screen.queryByTestId("teacher-upgrade-warning");
+    expect(warning).not.toBeInTheDocument();
+  });
+
+  it("displays a message, and no form, if the user is already a teacher", async () => {
+    setupTest({ role: "TEACHER" });
+    const message = await screen.findByTestId("already-teacher-account");
+    expect(message).toHaveTextContent("You already have a teacher account");
+    const form = screen.queryByRole("form");
+    expect(form).not.toBeInTheDocument();
+  });
+
+  it("displays a message, and no form, if the user already has a pending request", async () => {
+    const modifyUser = (user: typeof mockUser) =>
```

Comment:
Why not `RegisteredUserDTO` instead?

Reply:
The same goes for the type of `modifyUser` in `setupTest`.

Reply:
And in any usage of `typeof mockUser` in this file.

Reply:
Yeah makes sense, mockUser previously wasn't typed to RegisteredUserDTO but I changed that in this ticket